[Android] Connecting and cancelling connection to device with autoconnect leads to GATT Failure

Connecting with autoconnect on android, after several connection cancellations, fails immediately with a GATT Failure.

## Steps to reproduce

1. var cts = new CancellationTokenSource(1000);

2. await adapter.ConnectToKnownDeviceAsync(pairedDeviceId, new ConnectParameters(true), cts.Token);

3. Repeat 

## Expected behavior
App should continue cancelling and re-attempting to connect to device ad eternum

## Actual behavior
Exception is thrown with GATT Failure after several tries.

### Crashlog
```
04-23 21:10:33.518 E/mono-rt (15084): [ERROR] FATAL UNHANDLED EXCEPTION: Plugin.BLE.Abstractions.Exceptions.DeviceConnectionException: GattCallback error: Failure
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at Plugin.BLE.Abstractions.Utils.TaskBuilder+<FromEvent>d__1`3[TReturn,TEventHandler,TRejectHandler].MoveNext () [0x00121] in C:\Projects\xamarin-bluetooth-le\Source\Plugin.BLE.Abstractions\Utils\TaskBuilder.cs:51 
04-23 21:10:33.518 E/mono-rt (15084): --- End of stack trace from previous location where exception was thrown ---
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at Plugin.BLE.Abstractions.AdapterBase+<ConnectToDeviceAsync>d__43.MoveNext () [0x00184] in C:\Projects\xamarin-bluetooth-le\Source\Plugin.BLE.Abstractions\AdapterBase.cs:100 
04-23 21:10:33.518 E/mono-rt (15084): --- End of stack trace from previous location where exception was thrown ---
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at Plugin.BLE.Android.Adapter+<ConnectToKnownDeviceAsync>d__21.MoveNext () [0x00094] in C:\Projects\xamarin-bluetooth-le\Source\Plugin.BLE.Android\Adapter.cs:211 
04-23 21:10:33.518 E/mono-rt (15084): --- End of stack trace from previous location where exception was thrown ---
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
04-23 21:10:33.518 E/mono-rt (15084):   at BLEConnectLoopFail.MainActivity+<DoBluetoothConnectLoop>d__1.MoveNext () [0x000b1] in C:\Projects\xamarin-bluetooth-le\Source\BLEConnectLoopFail\MainActivity.cs:40 
```

### Probable cause of error
```
04-23 21:10:23.998 29113 29266 E bt_att  : GATT_Register: can't Register GATT client, MAX client reached!
04-23 21:10:23.998 29113 29266 E bt_btif : Register with GATT stack failed.
04-23 21:10:23.998 29113 29266 E bt_att  : GATT_Register: can't Register GATT client, MAX client reached!
04-23 21:10:23.998 29113 29266 E bt_btif : Register with GATT stack failed.
```

## Configuration

**Version of the Plugin:** e.g. 1.3.0, master

**Platform:** Android 7.1

**Device:** Xperia X Compact

--- a/Source/Plugin.BLE.Android/Adapter.cs
+++ b/Source/Plugin.BLE.Android/Adapter.cs
@@ -143,7 +143,7 @@ namespace Plugin.BLE.Android
         protected override Task ConnectToDeviceNativeAsync(IDevice device, ConnectParameters connectParameters,
             CancellationToken cancellationToken)
         {
-            ((Device)device).Connect(connectParameters);
+            ((Device)device).Connect(connectParameters, cancellationToken);
             return Task.CompletedTask;
         }
 

--- a/Source/Plugin.BLE.Android/Device.cs
+++ b/Source/Plugin.BLE.Android/Device.cs
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 using Android.App;
 using Android.OS;
@@ -77,30 +78,29 @@ namespace Plugin.BLE.Android
                 unsubscribeReject: handler => _gattCallback.ConnectionInterrupted -= handler);
         }
 
-        public void Connect(ConnectParameters connectParameters)
+        public void Connect(ConnectParameters connectParameters, CancellationToken cancellationToken)
         {
             IsOperationRequested = true;
             IsAutoConnectRequested = connectParameters.AutoConnect;
 
-            if (connectParameters.ForceBleTransport)
-            {
-                ConnectToGattForceBleTransportAPI(connectParameters.AutoConnect);
-            }
-            else
-            {
-                /*_gatt = */
-                BluetoothDevice.ConnectGatt(Application.Context, connectParameters.AutoConnect, _gattCallback);
+            var gatt = connectParameters.ForceBleTransport
+                ? ConnectToGattForceBleTransportAPI(connectParameters.AutoConnect)
+                : BluetoothDevice.ConnectGatt(Application.Context, connectParameters.AutoConnect, _gattCallback);
+
+            if (IsAutoConnectRequested)
+            {   // Close GATT to unregister device connection when connection request is canceled
+                cancellationToken.Register(() => gatt.Close());
             }
         }
 
-        private void ConnectToGattForceBleTransportAPI(bool autoconnect)
+        private BluetoothGatt ConnectToGattForceBleTransportAPI(bool autoconnect)
         {
             //This parameter is present from API 18 but only public from API 23
             //So reflection is used before API 23
             if (Build.VERSION.SdkInt < BuildVersionCodes.Lollipop)
             {
                 //no transport mode before lollipop, it will probably not work... gattCallBackError 133 again alas
-                BluetoothDevice.ConnectGatt(Application.Context, autoconnect, _gattCallback);
+                return BluetoothDevice.ConnectGatt(Application.Context, autoconnect, _gattCallback);
             }
             else if (Build.VERSION.SdkInt < BuildVersionCodes.M)
             {
@@ -111,13 +111,12 @@ namespace Plugin.BLE.Android
                                 Java.Lang.Integer.Type});
 
                 var transport = BluetoothDevice.Class.GetDeclaredField("TRANSPORT_LE").GetInt(null); // LE = 2, BREDR = 1, AUTO = 0
-                m.Invoke(BluetoothDevice, Application.Context, false, _gattCallback, transport);
+                return (BluetoothGatt)m.Invoke(BluetoothDevice, Application.Context, false, _gattCallback, transport);
             }
             else
             {
-                BluetoothDevice.ConnectGatt(Application.Context, autoconnect, _gattCallback, BluetoothTransports.Le);
+                return BluetoothDevice.ConnectGatt(Application.Context, autoconnect, _gattCallback, BluetoothTransports.Le);
             }
-
         }
 
         /// <summary>

--- a/Source/Plugin.BLE.Android/Device.cs
+++ b/Source/Plugin.BLE.Android/Device.cs
@@ -111,7 +111,7 @@ namespace Plugin.BLE.Android
                                 Java.Lang.Integer.Type});
 
                 var transport = BluetoothDevice.Class.GetDeclaredField("TRANSPORT_LE").GetInt(null); // LE = 2, BREDR = 1, AUTO = 0
-                return (BluetoothGatt)m.Invoke(BluetoothDevice, Application.Context, false, _gattCallback, transport);
+                return (BluetoothGatt)m.Invoke(BluetoothDevice, Application.Context, autoconnect, _gattCallback, transport);
             }
             else
             {

